js/impress.js

```diff
@@ -222,7 +174,9 @@
                 init: empty,
                 goto: empty,
                 prev: empty,
-                next: empty
+                next: empty,
+                tear: empty,
+                lib: {}
```

Comment:
I'm not sure if it's wise exposing all the internals of impress.js for consumers unless there's a reason to do so. The more we expose, the more we're required to support.

Reply:
lib (together with other API methods) are exposed to plugins, therefore ARE our public API.

Libs are not only meant for internal use, but to anyone who wants to build a plugin. So in such case they become a part of impress.js API.
Yes, this means we need to support these and maintain them. Theoretically it will make impress.js 'harder' to change later (to keep backwards compatibility), but it's a part of the process of making impress.js more modular and extensible. We cannot change impress.js if we don't change something ;)

Reply:
Cool then. Perhaps in the future, we'll be able to see the patterns clearer and that will allow us to make better decisions on what to expose and what not.

Reply:
tear() is a "public" API method, partner to init(). I will add it to DOCUMENTATION.md in this PR, thanks for pointing that out.

lib is "internal" in the sense that it is intended for plugins. But as you say, anybody could write plugins. My own idea here was that:
* As for documentation, README files and source code comments are sufficient. (Of course, we could adopt something like jsdoc if we want.)
*  As for public/private, I was thinking since we are an open source project, that the library API have similar guarantee as the Linux kernel: If we need to change them we will. If you don't want your plugin to break, you can contribute it to the upstream repo - with tests ;-) - and it won't break. If you have your own 3rd party plugin, make sure to check version of impress.js and *hopefully* the API only breaks between versions, not within.

Reply:
The second point is interesting, however, if we [restrict the scope](https://medium.com/@fagnerbrack/the-strictness-principle-9997e483cafb) by default we'll avoid unintended breakage and issue reports that will surely come. We can experiment with that approach, though, and see if it works for impress.js. Unfortunately, Github itself doesn't help when we try to apply Linux principles such as [Linux-like forking mechanisms](http://blog.ffwll.ch/2017/08/github-why-cant-host-the-kernel.html)...